Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2021, Sahan Fernando <sahan.h.fernando@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Graphics/VirtIOGPU/VirtIOGPU.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#define DEVICE_EVENTS_READ 0x0
+#define DEVICE_EVENTS_CLEAR 0x4
+#define DEVICE_NUM_SCANOUTS 0x8
+
+namespace Kernel::Graphics {
+
+VirtIOGPU::VirtIOGPU(PCI::Address address)
+    : VirtIODevice(address, "VirtIOGPU")
+    , m_scratch_space(MM.allocate_contiguous_kernel_region(8 * PAGE_SIZE, "VirtGPU Scratch Space", Region::Access::Read | Region::Access::Write))
+{
+    VERIFY(!!m_scratch_space);
+    dbgln("Gonna configure now");
+    if (auto cfg = get_config(ConfigurationType::Device)) {
+        m_device_configuration = cfg;
+        bool success = negotiate_features([&](u64 supported_features) {
+            u64 negotiated = 0;
+            if (is_feature_set(supported_features, VIRTIO_GPU_F_VIRGL))
+                dbgln("VirtIOGPU: VIRGL is not yet supported!");
+            if (is_feature_set(supported_features, VIRTIO_GPU_F_EDID))
+                dbgln("VirtIOGPU: EDID is not yet supported!");
+            return negotiated;
+        });
+        if (success) {
+            read_config_atomic([&]() {
+                m_num_scanouts = config_read32(*cfg, 0x8);
+            });
+            dbgln("VirtIOGPU: num_scanouts: {}", m_num_scanouts);
+            success = setup_queues(2); // CONTROLQ + CURSORQ
+        }
+        if (success) {
+            finish_init();
+        }
+        // 1. Get display information using VIRTIO_GPU_CMD_GET_DISPLAY_INFO
```

Comment:
what happens on `!success`?

Reply:
Changed this to VERIFY success (because it's almost certainly a bug if the virtio display doesn't initialize).